### PR TITLE
roachtest: UT for test selection

### LINKED ISSUE #129786: Unit Test for roachtest test selector
The test selection is used in the teamcity nightly runs. This is based on a snowflake query and a series of logic which is critical code. So, having unit test coverage for the code is very important to ensure its integrity.

Jira issue: CRDB-41709

### PR DESCRIPTION
**Pull request by nameisbhaskar**
This PR has the unit tests for the test selection changes. The tests the selector and the main code. This does not validate the snowflake query, but, simulates the response for other tests.

With this change, we have a coverage of 96% for the test selection.

Fixes: #129786
Epic: None

### CODE REVIEW

**Review by DarrylWong — COMMENTED** [review]
I haven't taken a deep look at this PR yet, but wrt #130330, it's good that it caught that bug, but the fact it was just an occasional flake makes me think we should add a bit more. We should have all the parts to do a full simulation (in addition to these hard coded edge cases).

i.e. generate some random starting state, run test selection, randomly pass/fail/preempt some tests, mock writing back the data, then repeat over N days. You can then check invariants at each step, i.e. preempted test is selected the next day, all tests have been run in last 4 days, etc.

Wdyt? Could also be a followup, since it might be a non trivial amount of code.

**Comment by nameisbhaskar** [comment]
> I haven't taken a deep look at this PR yet, but wrt #130330, it's good that it caught that bug, but the fact it was just an occasional flake makes me think we should add a bit more. We should have all the parts to do a full simulation (in addition to these hard coded edge cases).
> 
> i.e. generate some random starting state, run test selection, randomly pass/fail/preempt some tests, mock writing back the data, then repeat over N days. You can then check invariants at each step, i.e. preempted test is selected the next day, all tests have been run in last 4 days, etc.
> 
> Wdyt? Could also be a followup, since it might be a non trivial amount of code.

The current tests covers all known scenarios. But, randomised test is a good idea to generate Fuzzy kind of test cases - I will check how this can be achieved.

**Comment by nameisbhaskar** [comment]
@DarrylWong randomised test is making it complex for the assertion, but it should be doable. But, I am thinking of raising another PR later for that. Please do let me know if hat sound ok to you.

**Comment by bitomukesh** [comment]
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>Testing</b> - Add Unit Tests for Test Selection
</td>
<td width="70%">


<div>

-&nbsp;[`main_test.go`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-c6c7cdbdba90963e2083b43e3c86c43fe66dc304e198ad000af8b4f3cd3c2fb9) - Add unit tests for updateSpecForSelectiveTests function
<div>

-&nbsp;[`selector_test.go`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-7b3cf294882211e489dfc97da1e1e1376336f3b0bcaec6aabdf76b346715ce30) - Add comprehensive unit tests for test selection logic
<div>

-&nbsp;[`selector.go`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-8aa29de290677c76ceff9ad9fe54d589c3b4e33ab5e217da2c998636215015eb) - Minor modifications to support unit testing
<div>

-&nbsp;[`BUILD.bazel`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-47dc5ebecef00de3a18b37875f14acb7b51b07000dcf29d33e064d6ec9830419) - Update build configuration to include new test file
</div>
</td>
</tr>

<tr>
<td width="30%">
    <b>Other Improvements</b> - Dependency and Build Updates
</td>
<td width="70%">


<div>

-&nbsp;[`go.mod`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6) - Add github.com/DATA-DOG/go-sqlmock dependency
<div>

-&nbsp;[`BUILD.bazel`](https://github.com/cockroachdb/cockroach/pull/130079/files#diff-77c6e873c0e929f899a421c24a8dc11059db174aff0d501859ed00d8b419e10a) - Update build configuration to include new dependencies
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

File: pkg/cmd/roachtest/main_test.go [hunk 1]
```diff
@@ -98,3 +105,127 @@ func TestSampleSpecs(t *testing.T) {
...
+	ctx := context.Background()
+	var mock sqlmock.Sqlmock
+	var db *gosql.DB
+	var err error
+	_ = os.Setenv("SFUSER", "dummy_user")
+	_ = os.Setenv("SFPASSWORD", "dummy_password")
+	testselector.SqlConnectorFunc = func(_, _ string) (*gosql.DB, error) {
+		return db, err
+	}
+	t.Run("expect CategoriseTests to fail", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/main_test.go:119]: I will change the test name to "expect CategoriseTests to fail which causes fall back to run all tests". Also, will add comment.

File: pkg/cmd/roachtest/main_test.go [hunk 2]
```diff
@@ -98,3 +105,127 @@ func TestSampleSpecs(t *testing.T) {
...
+// all tests which contains "_skipped" are skipped in teh spec
+// all tests which contains "_selected" are selected="yes"
+// all tests which contains "_skip_selector" are skipped based on the criteria
+// all tests which contains "_missing" are missing in the test spec
+// all tests which contains "_success" are selected="no"
+// all tests which contains "new_test" are present in specs, but missing in snowflake rows
+// all tests which contains "_randomized" are set in spec as Randomized=true
+func getMockData() ([]registry.TestSpec, *sqlmock.Rows) {
+	specs := []registry.TestSpec{
+		{Name: "t2_selected"},
```
[the remark this answers is not in the document]
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/main_test.go:180]: Sure, I will add a comments. The entries have been manually shuffled around and the some names were added later, so, is the mismatch in names.

**Comment by nameisbhaskar** [comment]
Thanks @DarrylWong  for your review. I have addressed the comments. Please do take a look.

File: pkg/cmd/roachtest/testselector/selector_test.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			FirstRunOn:  2,
+			LastRunOn:   3,
+			Cloud:       spec.AWS,
+			Suite:       "unit test",
+		})
+		require.Nil(t, tds)
+		require.NotNil(t, err)
+		require.Equal(t, "failed to execute query", err.Error())
+	})
+	t.Run("expect no failure", func(t *testing.T) {
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/testselector/selector_test.go:82]: nit: a more descriptive test name might mention `the sequence of response list must be maintained.`, which as you noted below is what we're really testing here.

File: pkg/cmd/roachtest/testselector/selector_test.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+func TestNewDefaultSelectTestsReq(t *testing.T) {
+	req := NewDefaultSelectTestsReq(spec.Azure, "ut suite")
+	require.Equal(t, spec.Azure, req.Cloud)
+	require.Equal(t, "ut suite", req.Suite)
+	require.Equal(t, defaultForPastDays, req.ForPastDays)
+	require.Equal(t, defaultFirstRunOn, req.FirstRunOn)
+	require.Equal(t, defaultLastRunOn, req.LastRunOn)
+}
+
+func Test_getSFCreds(t *testing.T) {
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/testselector/selector_test.go:133]: Similar comment to above, these tests seem to be already covered by `TestCategoriseTests`? Not sure how much extra coverage we get.

File: pkg/cmd/roachtest/main_test.go [hunk 5]
```diff
@@ -98,3 +105,136 @@ func TestSampleSpecs(t *testing.T) {
...
+// all tests which contains "_selected" are selected="yes"
+// all tests which contains "_preempted" are last_failure_is_preempt="yes"
+// all tests which contains "_skip_selector" are skipped based on the criteria
+// all tests which contains "_missing" are missing in the test spec
+// all tests which contains "_success" are selected="no"
+// all tests which contains "new_test" are present in specs, but missing in snowflake rows
+// all tests which contains "_randomized" are set in spec as Randomized=true
+// Other than the above conventions, the name of the test is randomly put to represent random names for tests.
+// the test names are also shuffled to ensure that the test is more close to real data.
+func getMockData() ([]registry.TestSpec, *sqlmock.Rows) {
```
Comment (DarrylWong) [inline pkg/cmd/roachtest/main_test.go:182]: nit: maybe `getTestSelectionMockData`? Since this is in `main_test.go` lets make it specific.

**Review by DarrylWong — COMMENTED** [review]
This mostly looks good to me. My remaining comments are mostly just noting that some tests seem redundant to me.

File: pkg/cmd/roachtest/testselector/selector_test.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		dsn, err := getDSN()
+		require.Equal(t, "snowflake", driverName)
+		require.Equal(t, dsn, dataSourceName)
+		return db, err
+	}
+	var err error
+	t.Run("expect Prepare to fail", func(t *testing.T) {
+		db, mock, err = sqlmock.New()
+		require.Nil(t, err)
+		mock.ExpectPrepare(regexp.QuoteMeta(PreparedQuery)).WillReturnError(fmt.Errorf("failed to prepare"))
```
[the remark this answers is not in the document]
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/testselector/selector_test.go:60]: This is to test that we are returning the error correctly from the code.

File: pkg/cmd/roachtest/testselector/selector_test.go (hunk 4 above)
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/testselector/selector_test.go:133]: This is for the coverage of all error scenarios and ensure that we are returning the errors.

**Comment by nameisbhaskar** [comment]
Thanks @DarrylWong ! I have replied to your comments. Please do let me know if it sounds good.

**Review by DarrylWong — APPROVED** [review]

**Comment by nameisbhaskar** [comment]
Thanks @DarrylWong for reviewing this.

**Comment by nameisbhaskar** [comment]
bors r=@DarrylWong
